### ISSUE TITLE
Option to bypass the generation of GPOS features

### ISSUE THREAD

**Issue by miguelsousa** [googlei18n/fontmake#255]
I'd like to add an option to *fontmake* for it to bypass the generation of GPOS features (`kern`, `mark`, `mkmk`) and let it use FEA files that are already in place instead.

Currently I'm using a file that contain the classes below, which are called with the `--kern-writer-module` and `--mark-writer-module` options, but I imagine that there's a simpler way to implement the desired functionality. Can someone provide a suggestion on how to go about it? Thanks!

```python
class KernFeatureWriter(object):

	def __init__(self, font):
		pass

	def write(self):
		return ""


class MarkFeatureWriter(object):

	def __init__(self, font, anchorList, mkmkAnchorList=(), ligaAnchorList=()):
		pass

	def write(self, doMark=True, doMkmk=True):
		return ""
```

**Comment by brawer** [comment]
Wouldn’t it be best if UFOs were “self-contained” so that a given input file could be sent to any compiler and it always produces the same output? Such “self-contained UFOs” would be simpler, and there’d be fewer places where the process can go wrong. For example, instead of telling the compiler to ignore [kerning.plist](http://unifiedfontobject.org/versions/ufo3/kerning.plist/), why simply not omitting that file from the input UFO file?

Admittedly, I don’t understand the reason why custom kern/mark feature writers were added to fontmake. From a simplicity perspective, it seems better to have less options. Obviously I’m missing something; can anyone enlighten me?

**Comment by miguelsousa** [comment]
Removing the kerning.plist file from the UFO in not an option, because that (plus groups.plist) contains the kerning data in raw format, which is needed for editing, interpolation, etc.
The thing is, there's more than one way of converting that data to FEA format. The same can be said about the anchor data, which gets assembled as `mark` and `mkmk` FEA code. We have [our own scripts](https://github.com/adobe-type-tools/python-modules) to do all this work, but they're not setup to hook up to fontmake, and we really don't want nor need to be regenerating the FEA inputs every time we build the fonts. So we only do it when it's necessary, and we keep the files under version control. We also have projects that require manual adjustments to the FEA files, like contextual kerning, which currently cannot be made on-the-fly.
As you can see there are several reasons for wanting to build fonts with pre-generated GPOS FEA files, which, in most cases, were made from the data that lives inside the UFO. They're just not remade every single time the font is built.

**Comment by brawer** [comment]
> there's more than one way of converting that data

Yes, exactly that’s my concern. We won’t get streamlined workflows when the same piece of data can be converted to output in different ways. (Unless there’s a specification to clearly decide which way is correct).

But thanks for your patient explanation. It sounds like you’d like to have something in the exact format of `kerning.plist`, but without the associated semantics. Have you considered putting your files into the [data directory](http://unifiedfontobject.org/versions/ufo3/data/) inside the UFO, for example as `data/raw-kerning.plist` or so?

If that wasn’t an option for some reason (curious to hear more), what would you think of defining a custom property in [lib.plist](http://unifiedfontobject.org/versions/ufo3/lib.plist/) that instructs the compiler to ignore `kerning.plist` etc.?

Basically, I’m looking for a way to making builds “reproducible”, meaning it should be clearly defined how to convert some piece of input to output. If we add custom compiler options that aren’t part of the UFO file itself, the build process becomes harder to reproduce because it’s difficult to know when to pass what flag.

**Comment by miguelsousa** [comment]
@brawer I feel that you're overthinking and it's probably my fault. Let me restate the situation:
• I have a UFO font that is ready to be generated into an OTF.
• The [`features.fea` file](http://unifiedfontobject.org/versions/ufo3/features.fea/) inside that UFO contains GSUB as well as GPOS features, namely `kern`, `mark` and `mkmk`.
• When I run this UFO thru **fontmake** using no options (`fontmake -u font.ufo`) it generates its own `kern`, `mark` and `mkmk` features and appends them to the already existing FEA code, resulting in two sets of GPOS features.
• To prevent **fontmake** from appending its generated GPOS features, I use the classes described in the original comment by saving them in a file, and running this command instead, `fontmake -u font.ufo --kern-writer-module MyGPOSModule --mark-writer-module MyGPOSModule`, which essentially results in the appending of an empty string, therefore leaving the original code in `features.fea` intact.

That said, I'm looking for a way of telling **fontmake**, *Thanks, but I don't want the GPOS feature code that you generate at build time*. The solution I have in place works fine, but I feel that there must be a simpler way of implementing this, likely as an option, which could benefit other users of **fontmake**.

**Comment by anthrotype** [comment]
Hey Miguel, this should already be the case. If it's not, then it's likely a ufo2ft bug.

ufo2ft uses a simple regex to check for the presence of kern, mark or mkmk features and by default it is configured to not overwrite:
https://github.com/googlei18n/ufo2ft/blob/5507abd8a765d44587502cbdc16d441e3cd28bef/Lib/ufo2ft/makeotfParts.py#L78

**Comment by miguelsousa** [comment]
@anthrotype didn't know about that, thanks for pointing it out.
I see that I've oversimplified my example. The UFO's `features.fea` actually contains only an `include()` statement, so that sniffing code won't find them 😞

**Comment by anthrotype** [comment]
could you file the issue on ufo2ft repo?
I wonder if the feaLib parser could in ufo2ft to find out whether kern/mark/mkmk have already been defined/included.

**Comment by miguelsousa** [comment]
Will do

**Comment by anthrotype** [comment]
> Admittedly, I don’t understand the reason why custom kern/mark feature writers were added to fontmake.

I think the idea was to allow one to plug into fontmake one's own custom feature writers, though in practice (from Miguel's example, and Adam's "feaLab") most seem to be using these options just to skip the feature writing altogether.

I haven't tried to use those options yet, but if you think they are unnecessary we can deprecate/remove them.

Another reason why people would like a `--no-gpos` option (or something equivalent like the above mentioned dummy feature writers) is that varLib sometimes can't successfully merge the varfonts' GPOS tables.

**Comment by twardoch** [comment]
@brawer The reason why fontmake has a converter from UFO structures for kerning and mark positioning is that in some cases, it's reasonable to have the higher-level structures which are more editable (kerning.plist etc.), but the "kern" and "mark" features are much more capable than what the FontLab Studio 5 simplified model withthe "top"/"_top" anchor names, or the UFO group kerning model can define. 

In "kern",  you can have contextual kerning and multiple lookups, in "mark" you can also have contextual positionings or such that e.g. modify the base glyph width when placing very wide marks. 

kerning.plist is merely a simple partial template for GPOS. There is a need for competing logics as to how to generate GPOS from UFO, because UFO is not a full description of OpenType, just a higher-level subset that describes the most common scenarios. UFO never had been particularly optimized for non-Latin (which is e.g. why horizontal LTR kerning has its dedicated structure bukerning in other directions does not). 

Out of OpenType's 8 glyph definition formats (glyf, CFF, CFF2, SVG, bdat/EBDT, sbix, CBDT), UFO only provides higher-level structures for the first two. It is a wrapper that works reasonably well for Latin+Cyrillic+Greek fonts with monochrome PostScript outlines, but less so for anything else. This probably is a good thing, otherwise it would have been far too complicated.

**Comment by twardoch** [comment]
Ps. I have used fontmake with a FontLab in-house kern feature writer, which will likely be included in FontLab VI.

**Comment by brawer** [comment]
Have a look at https://github.com/googlei18n/ufo2ft/pull/110. This might be controversial, but I didn’t see any reasonable way to handle https://github.com/googlei18n/ufo2ft/issues/108 while also keeping things fully configurable. (And I’m still not convinced why making things externally configurable would be a good thing, but I might have misunderstood something). Feel free to tear the pull request apart, ideally at the review thread for the pull request.

**Comment by miguelsousa** [comment]
@brawer the PR looks good, thank you. I'll try it out and report back shortly.

**Comment by miguelsousa** [comment]
@brawer I've used your changes to build one of the OTFs at https://github.com/adobe-type-tools/otf-making-benchmark and all went well. As far as I'm concerned it can be merged.
Thanks again for considering this enhancement.

**Comment by brawer** [comment]
OK, merged.
